### PR TITLE
read LSP7/8 using LSP4TokenType

### PR DESCRIPTION
**Pull request by biancabuzea200**

### CODE REVIEW

File: docs/learn/smart-contract-developers/readLSP7andLSP8.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+# Read LSP7/8 Tokens
+
+## Background
+
+Currently LSP7 and LSP8 can both be used as NFTs. LSP7 is useful for mass-digital NFTs, where individual items are not unique, whereas LSP8 is mainly used for phygitals, or NFTs with unique properties per item.
```
Comment (fhildeb) [inline docs/learn/smart-contract-developers/readLSP7andLSP8.md:13]:
```suggestion
The [LSP7](../../standards/tokens/LSP7-Digital-Asset.md) and [LSP8](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md) assets can both be used as NFTs. LSP7 is useful for mass-digital NFTs, where individual items are not unique, whereas LSP8 is mainly used for phygitals, or NFTs with unique properties per item.
```

File: docs/learn/smart-contract-developers/readLSP7andLSP8.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+Previously, an interface could not differentiate between LSP7 being a token or an NFT. [LSP4TokenType data key](https://github.com/lukso-network/LIPs/blob/main/LSPs/LSP-4-DigitalAsset-Metadata.md?plain=1) allows to define the type, in a smart contract readable form.
```
Comment (fhildeb) [inline docs/learn/smart-contract-developers/readLSP7andLSP8.md:15]:
```suggestion
To detect the fungibility of an asset, the 📄 [LSP4 Digital Asset Metadata](../../standards/tokens/LSP4-Digital-Asset-Metadata.md) standard defines the `LSP4TokenType`. This data key can be retrieved from the smart contracts 🗂️ [ERC725Y](../../standards/lsp-background/erc725.md#erc725y-generic-data-keyvalue-store) storage.
```

File: docs/learn/smart-contract-developers/readLSP7andLSP8.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+{
+  "name": "LSP4TokenType",
+  "key": "0xe0261fa95db2eb3b5439bd033cda66d56b96f92f243a8228fd87550ed7bdfdb3", // kecca256 hash of the word « LSP4TokenType »
+  "keyType": "Singleton",
+  "valueType": "uint256",
+  "valueContent": "Number"
+}
+```
+
+This value is not changeable, and is set during the token's initialization.
```
Comment (fhildeb) [inline docs/learn/smart-contract-developers/readLSP7andLSP8.md:27]:
```suggestion
The `LSP4TokenType` is not changeable and set during the token's initialization.
```

File: docs/learn/smart-contract-developers/readLSP7andLSP8.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+|  `2`  | `Collection` | Only valid for LSP8. The `LSP4Metadata` represents the information of a the collection, and each individual token ID represents its own NFT and MUST have its own metadata set using `LSP8TokenIdType` and `LSP8MetadataTokenURI`. [See LSP8 for details](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md).                                                                                                                                                                                                                                                         |
+
+## Implementation
+
+When creating smart contracts representing digital assets on LUKSO, you will need to specify the token type and data keys for the 📄 [LSP4 Digital Asset Metadata](../../standards/tokens/LSP4-Digital-Asset-Metadata) that will be stored in the 🗂️ [ERC725Y](../../standards/lsp-background/erc725.md#erc725y-generic-data-keyvalue-store) storage of the Digital Asset. There are three different token types:
```
Comment (fhildeb) [inline docs/learn/smart-contract-developers/readLSP7andLSP8.md:37]:
```suggestion
When creating digital assets using [LSP7](../../standards/tokens/LSP7-Digital-Asset.md) or [LSP8](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md), the type and data keys of the 📄 [LSP4 Digital Asset Metadata](../../standards/tokens/LSP4-Digital-Asset-Metadata.md) have to be defined within the smart contract like the following:
```

This sentence seemed like a duplication of the sections above. The snippet you show does not specifically write into the ERC725Y. Just when inheriting from LSP7/LSP8 standards, this will be added into ERC725Y storage. 

Therefore, I would make the sentence more straightforward. What I also noticed: sometimes you use `.md` on links, and sometimes you leave it out. We should stick to the md-based format on internal links.

File: docs/learn/smart-contract-developers/readLSP7andLSP8.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+bytes32 constant _LSP4_TOKEN_TYPE_DATA_KEY = 0xe0261fa95db2eb3b5439bd033cda66d56b96f92f243a8228fd87550ed7bdfdb3; // kecca256 hash of the word `LSP4TokenType`
+
+enum TokenType {
+    TOKEN,     // `0` = Token
+    NFT,       // `1` = NFT
+    COLLECTION // `2` = Collection
+}
+```
+
+After defining the token type of the asset you can create a custom [LSP7 Digital Asset Collection](../../standards/tokens/LSP7-Digital-Asset.md) or [LSP8 Identfiable Digital Asset Collection](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md).
```
Comment (fhildeb) [inline docs/learn/smart-contract-developers/readLSP7andLSP8.md:52]:
```suggestion
After defining the token type of the asset, you can create a custom [LSP7 Digital Asset Collection](../../standards/tokens/LSP7-Digital-Asset.md) or [LSP8 Identifiable Digital Asset Collection](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md). During deployment, the token type is then written to the 🗂️ [ERC725Y](../../standards/lsp-background/erc725.md#erc725y-generic-data-keyvalue-store) storage of the smart contract.
```

**Review by fhildeb — COMMENTED** [review]
Good addition. Left some grammar fixes and suggestions to improve clarity. 😊

**Review by fhildeb — APPROVED** [review]

File: docs/learn/smart-contract-developers/readLSP7andLSP8.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+# Read LSP7/8 Tokens
+
+## Background
+
+[LSP7](../../standards/tokens/LSP7-Digital-Asset.md) and [LSP8](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md) can be both used as NFTs. LSP7 is useful for mass-digital NFTs, where individual items are not unique, whereas LSP8 is mainly used for phygitals, or NFTs with unique properties per item.
```
Comment (Hugoo) [inline docs/learn/smart-contract-developers/readLSP7andLSP8.md:13]: What does "mass-digital" NFT mean?

**Comment by Hugoo** [comment]
Is there a reason for using a different case for this file name?

`readLSP7andLSP8` and not `read-lsp7-and-lsp8` or similar?

What would be the reason to have different case for that file?

image

File: docs/learn/smart-contract-developers/readLSP7andLSP8.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+|  `2`  | `Collection` | Only valid for LSP8. The `LSP4Metadata` represents the information of a the collection, and each individual token ID represents its own NFT and MUST have its own metadata set using `LSP8TokenIdType` and `LSP8MetadataTokenURI`. [See LSP8 for details](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md).                                                                                                                                                                                                                                                         |
+
+## Implementation
+
+When creating digital assets using [LSP7](../../standards/tokens/LSP7-Digital-Asset.md) or [LSP8](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md), the type and data keys of the 📄 [LSP4 Digital Asset Metadata](../../standards/tokens/LSP4-Digital-Asset-Metadata.md) have to be defined within the smart contract like the following:
+
+```solidity title="contracts/TokenTypes.sol"
```
Comment (Hugoo) [inline docs/learn/smart-contract-developers/readLSP7andLSP8.md:39]: What does this filename refer to" `contracts/TokenTypes.sol`

File: docs/learn/smart-contract-developers/readLSP7andLSP8.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+bytes32 constant _LSP4_TOKEN_TYPE_DATA_KEY = 0xe0261fa95db2eb3b5439bd033cda66d56b96f92f243a8228fd87550ed7bdfdb3; // kecca256 hash of the word `LSP4TokenType`
+
+enum TokenType {
+    TOKEN,     // `0` = Token
+    NFT,       // `1` = NFT
+    COLLECTION // `2` = Collection
+}
```
Comment (Hugoo) [inline docs/learn/smart-contract-developers/readLSP7andLSP8.md:49]: where is the reading part? As suggested by the title?

What is the goal of this article?

**Review by Hugoo — CHANGES_REQUESTED** [review]
I think a part is missing.

I don't see much related to "reading" the NFT?

It justs explains that there is a "LSP4TokenType" with 3 values and thats it.

Would be nice to have code sample to actually fetch this key and do something with it.

File: docs/learn/smart-contract-developers/readLSP7andLSP8.md (hunk 8 above)
Comment (biancabuzea200) [reply docs/learn/smart-contract-developers/readLSP7andLSP8.md]: This guide has been requested yesterday by Fabian. The reading refers to the interpretation of LSP7 and LSP8;  the guide should reflect the new additions introduced here https://github.com/lukso-network/LIPs/pull/250

File: docs/learn/smart-contract-developers/retrieve-token-type.md [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+---
+
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+# Retrieve Token Type
+
+## Background
+
+[LSP7](../../standards/tokens/LSP7-Digital-Asset.md) and [LSP8](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md) can be both used as NFTs. LSP7 is useful for NFTs where individual items are not unique, whereas LSP8 is mainly used for phygitals, or NFTs with unique properties per item.
```
Comment (CJ42) [inline docs/learn/smart-contract-developers/retrieve-token-type.md:13]: Suggesting to add:

```suggestion
[LSP7](../../standards/tokens/LSP7-Digital-Asset.md) and [LSP8](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md) can be both used as NFTs. LSP7 is useful for NFTs where individual items are not unique and to mint large quantity of NFTs at once, whereas LSP8 is mainly used for phygitals, or NFTs with unique properties per item.
```

File: docs/learn/smart-contract-developers/retrieve-token-type.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+To detect the fungibility of an asset, the 📄 [LSP4 Digital Asset Metadata](../../standards/tokens/LSP4-Digital-Asset-Metadata.md) standard defines the `LSP4TokenType`. This data key can be retrieved from the smart contracts 🗂️ [ERC725Y](../../standards/lsp-background/erc725.md#erc725y-generic-data-keyvalue-store) storage.
```
Comment (CJ42) [inline docs/learn/smart-contract-developers/retrieve-token-type.md:15]:
```suggestion
To detect the fungibility of an asset, the 📄 [LSP4 Digital Asset Metadata](../../standards/tokens/LSP4-Digital-Asset-Metadata.md) standard defines the `LSP4TokenType`. This data key can be retrieved from the 🗂️ [ERC725Y](../../standards/lsp-background/erc725.md#erc725y-generic-data-keyvalue-store) storage of the digital asset contract
```

File: docs/learn/smart-contract-developers/retrieve-token-type.md [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+|  `2`  | `Collection` | Only valid for LSP8. The `LSP4Metadata` represents the information of a the collection, and each individual token ID represents its own NFT and MUST have its own metadata set using `LSP8TokenIdType` and `LSP8MetadataTokenURI`. [See LSP8 for details](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md).                                                                                                                                                                                                                                                         |
+
+## Implementation
+
+When creating digital assets using [LSP7](../../standards/tokens/LSP7-Digital-Asset.md) or [LSP8](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md), the type and data keys of the 📄 [LSP4 Digital Asset Metadata](../../standards/tokens/LSP4-Digital-Asset-Metadata.md) have to be defined within the smart contract like the following:
```
Comment (CJ42) [inline docs/learn/smart-contract-developers/retrieve-token-type.md:37]: Suggest rewording:
M
```suggestion
When creating digital assets using [LSP7](../../standards/tokens/LSP7-Digital-Asset.md) or [LSP8](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md), the token type is defined under the `LSP4TokenType` data key. We can retrieve the value and type using `getData`
```

File: docs/learn/smart-contract-developers/retrieve-token-type.md [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+```solidity
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+bytes32 constant _LSP4_TOKEN_TYPE_DATA_KEY = 0xe0261fa95db2eb3b5439bd033cda66d56b96f92f243a8228fd87550ed7bdfdb3; // kecca256 hash of the word `LSP4TokenType`
```
Comment (CJ42) [inline docs/learn/smart-contract-developers/retrieve-token-type.md:43]:
```suggestion
bytes32 constant _LSP4_TOKEN_TYPE_DATA_KEY = 0xe0261fa95db2eb3b5439bd033cda66d56b96f92f243a8228fd87550ed7bdfdb3; // keccak256 hash of the word `LSP4TokenType`
```

File: docs/learn/smart-contract-developers/retrieve-token-type.md [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+|  `2`  | `Collection` | Only valid for LSP8. The `LSP4Metadata` represents the information of a the collection, and each individual token ID represents its own NFT and MUST have its own metadata set using `LSP8TokenIdType` and `LSP8MetadataTokenURI`. [See LSP8 for details](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md).                                                                                                                                                                                                                                                         |
+
+## Implementation
+
+When creating digital assets using [LSP7](../../standards/tokens/LSP7-Digital-Asset.md) or [LSP8](../../standards/tokens/LSP8-Identifiable-Digital-Asset.md), the type and data keys of the 📄 [LSP4 Digital Asset Metadata](../../standards/tokens/LSP4-Digital-Asset-Metadata.md) have to be defined within the smart contract like the following:
+
+```solidity
```
Comment (CJ42) [inline docs/learn/smart-contract-developers/retrieve-token-type.md:39]: I am not sure I understand why there is some Solidity code snippet here.

If the guide is to demonstrate how to read the tokenType, then maybe we could show how to do this in Javascript / Typescript? For dApp developers?
We can also show how to do it in Solidity, but I guess guides for dApp developers are more important?

**Review by CJ42 — CHANGES_REQUESTED** [review]
Added review comments

File: docs/learn/smart-contract-developers/retrieve-token-type.md (hunk 13 above)
Comment (biancabuzea200) [reply docs/learn/smart-contract-developers/retrieve-token-type.md:39]: right now this guide is under smart contract dev. I was thinking to leave this one here for SC dev, and maybe add another one for dApp developer (I think the Solidity part is also important). 
Wdyt?

**Comment by Hugoo** [comment]
note: lsp8 will change. lot this week so we need to be careful there.

**Review by Hugoo — APPROVED** [review]
im merging all these PRs and the SC team will go over the docs today and send fixes
